File: src/test/clojure/maunaloa/service/calculator/testblackscholes.clj
(ns maunaloa.service.calculator.testblackscholes
  (:require [maunaloa.service.calculator.BlackScholes :as B])
  (:use clojure.test))


(defn calc-diff [expected func spot x t sigma]
  (let [bs (func spot x t sigma)
        diff (Math/abs (- bs expected))]
    ;(prn func "option price: " bs ", sigma: " sigma ", diff: " diff)
    diff))

(deftest call-put-1 []
  (let [spot 100
        x 100
        t 0.5
        sigma 0.2]
    (is (<= (calc-diff 6.88 B/call-price spot x t sigma) 0.01) "Call atm 1")
    (is (<= (calc-diff 2.97 B/call-price spot x t (- sigma 0.15)) 0.01) "Call atm 2")
    (is (<= (calc-diff 13.78 B/call-price spot x t (+ sigma 0.25)) 0.03) "Call atm 3")
    (is (<= (calc-diff 21.94 B/call-price spot x t (+ sigma 0.55)) 0.03) "Call atm 4")
    (is (<= (calc-diff 4.66 B/put-price spot x t sigma) 0.241) "Put atm")
    (is (<= (calc-diff 0.73 B/put-price spot x t (- sigma 0.15)) 0.238) "Put atm 2")
    (is (<= (calc-diff 11.53 B/put-price spot x t (+ sigma 0.25)) 0.242) "Put atm 3")
    (is (<= (calc-diff 19.7 B/put-price spot x t (+ sigma 0.55)) 0.259) "Put atm 4")))


    (comment
(deftest call-put-2 []
  (let [spot 120
        x 100
        t 0.5
        sigma 0.2]
    (is (<= (calc-diff 6.88 B/call-price spot x t sigma) 0.01) "Call price 1")
    (is (<= (calc-diff 4.42 B/put-price spot x t sigma) 0.01) "Put price 1")))

(deftest call-put-3 []
  (let [spot 140
        x 100
        t 0.5
        sigma 0.2]
    (is (<= (calc-diff 6.88 B/call-price spot x t sigma) 0.01) "Call price 1")
    (is (<= (calc-diff 4.42 B/put-price spot x t sigma) 0.01) "Put price 1")))
  )

File: src/test/java/maunaloa/service/calculator/TestBlackScholes.java
package maunaloa.service.calculator;

import maunakea.financial.beans.CalculatedDerivativeBean;
import oahu.financial.OptionCalculator;
import oahu.financial.beans.DerivativeBean;
import oahu.financial.beans.StockBean;
import org.joda.time.DateMidnight;
import org.junit.Test;
import org.springframework.context.ApplicationContext;
import org.springframework.context.support.ClassPathXmlApplicationContext;

import java.util.Date;

import static org.junit.Assert.assertEquals;

//import org.junit.Before;

/**
 * Created with IntelliJ IDEA.
 * User: rcs
 * Date: 5/22/13
 * Time: 9:19 AM
 */
public class TestBlackScholes {

    private StockBean createStockBean(double spot) {

        return new StockBean(new Date(),0.0,0.0,0.0,spot,0);
    }
    /*
  asusupposfsddsfsd  @Before
    public void setUp() throws Exception {
    }

    @After
    public void tearDown() throws Exception {
    }
    //*/

    private OptionCalculator getCalculator() {
        ApplicationContext factory = new ClassPathXmlApplicationContext("test-maunaloa.xml");
        return factory.getBean("calculator",OptionCalculator.class);
    }

    private CalculatedDerivativeBean createDerivativeBean(StockBean parent,
                                                          int opType,
                                                          double x,
                                                          double buy,
                                                          double sell,
                                                          OptionCalculator calc) {
        Date dx = new DateMidnight().plusDays(183).toDate();

        return new CalculatedDerivativeBean(String.format("Test %d",opType),
                                            opType,
                                            x,
                                            buy,
                                            sell,
                                            dx,
                                            parent,
                                            calc);
    }

    @Test
    public void testDelta() {
        OptionCalculator calculator = getCalculator();

        CalculatedDerivativeBean bean =
        createDerivativeBean(createStockBean(100),
                             DerivativeBean.CALL,
                             100,
                             12,
                             14.5,
                             calculator);
        assertEquals("Call delta 1",0.6,bean.getDelta(),0.01);
    }

    @Test
    public void testStockPriceFor() {
        OptionCalculator calculator = getCalculator();

        CalculatedDerivativeBean bean =
                createDerivativeBean(createStockBean(100),
                        DerivativeBean.CALL,
                        100,
                        12,
                        14.5,
                        calculator);

        System.out.println(bean.getIvBuy());
        assertEquals("Stock price for call price 1",103.1,calculator.stockPriceFor(14,bean),0.1);
    }
}
